### PR TITLE
Error handler middleware

### PR DESCRIPTION
**Pull request by JoeChapman**
- requires a translat function argument
- accepts a logger option
- minor bump

### CODE REVIEW

File: lib/errors.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    if (err.code === 'NO_COOKIES') {
+      err.status = 403;
+      content.title = translate('errors.cookies-required.title');
+      content.message = translate('errors.cookies-required.message');
+    }
+
+    content.title = content.title || translate('errors.default.title');
+    content.message = content.message || translate('errors.default.message');
```
Comment (daniel-ac-martin) [inline lib/errors.js]: How useful is a default message? I wonder if in the future we should try to get something out out of the err object.

In lev-web we're currenltly doing something like this:
```js
    content.message = content.message || err.message || err || translate('errors.default.message');
```

Which isn't pretty but gives you some info when debugging.

I'm not sure what security would make of it.

Maybe it needs to be conditional on config.env?

File: lib/errors.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    if (logger && logger.error) {
+      logger.error(err.message || err.error, err);
+    }
+
+    res.render('error', {
+      error: err,
+      content: content,
+      /* eslint no-process-env:0 */
+      showStack: (process.env.NODE_ENV === 'development' || process.env.NODE_ENV === 'docker-compose'),
```
Comment (daniel-ac-martin) [inline lib/errors.js]: Should the "docker-compose" environment be a 'thing' in HOF?

Perhaps instead the middleware should take a 'debug' option?

Comment (JoeChapman) [inline lib/errors.js]: I'll like the sounds of a debug option.
I do foresee docker-compose being a commonality in HOF, at least in the example form

File: lib/errors.js (hunk 1 above)
Comment (JoeChapman) [inline lib/errors.js]: Security? I suppose it might expose something about our system. I don't know, I am happy to leave it as is, @easternbloc @joefitter ?

Comment (joefitter) [inline lib/errors.js]: 👍  I would go with checking the env before logging out unsanitised err object

File: lib/errors.js [hunk 3]
```diff
@@ -35,7 +35,7 @@ module.exports = (translate, options) => {
       error: err,
       content: content,
       /* eslint no-process-env:0 */
-      showStack: (process.env.NODE_ENV === 'development' || process.env.NODE_ENV === 'docker-compose'),
+      showStack: (options || {}).debug === true ? true : false,
```
Comment (daniel-ac-martin) [inline lib/errors.js]: It's possible there is a quirk in javascript I don't understand, but I reckoning you could delete everything after the first "true" (except the comma) on this line and it should mean exactly the same thing.

File: lib/errors.js (hunk 1 above)
Comment (daniel-ac-martin) [inline lib/errors.js]: Yes I believe that is the security concern, simply exposing info about the system unnecessarily.

I think it's good enough the way it is, but it might be nicer to have enhanced behaviour in debug mode.

File: lib/errors.js (hunk 3 above)
Comment (joefitter) [inline lib/errors.js]: agree

File: lib/errors.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+module.exports = (translate, options) => {
+
+  var logger = (options || {}).logger;
+
+  if (typeof translate !== 'function') {
+    throw new Error('Errorhandler needs a translate function');
+  }
+
+  return (err, req, res) => {
```
Comment (JoeChapman) [inline lib/errors.js]: I just noticed there is no `next` param. 
)

**Comment by JoeChapman** [comment]
@UKHomeOffice/hof

File: lib/errors.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
+'use strict';
+
+const getContent = (err, debug, translate) => {
+  var content = {};
```
Comment (easternbloc) [inline lib/errors.js]: if you're `const`ing and `let`ing perhaps stop `var`ing?

I think if we're gonna start forcing the es6 route we can stop with `var` altogether?

File: lib/errors.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+  if (err.code === 'SESSION_TIMEOUT') {
+    content.title = translate('errors.session.title');
```
Comment (easternbloc) [inline lib/errors.js]: not sure what i think of this. If feels quite frameworky. We are now saying you want a middleware you must have a `translate`. The idea of the cookie middleware was to keep it very light...

File: lib/errors.js (hunk 5 above)
Comment (JoeChapman) [inline lib/errors.js]: `const` is for things that won't change and `let` is for things that are intended to be scoped to a block, not a function

Comment (daniel-ac-martin) [inline lib/errors.js]: Arguably, one should avoid mutation in functional code but I don't think that level of discipline is normal in JavaScript.

File: lib/errors.js (hunk 6 above)
Comment (JoeChapman) [inline lib/errors.js]: However much we refrain from joining the framework club we are still using HOF as a tool to build a large number of forms, which IMO should be standardised to a point.
But, I am not fussy, perhaps it is too much to ask that a user always provides a translations function.

Comment (daniel-ac-martin) [inline lib/errors.js]: I can't see much of an alternative here. Unless we want to have the translations provided directly instead of the translation library?

At least we aren't prescribing which translation lib they use, and they can opt not to make use of the middleware is they choose.

File: lib/errors.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  if (debug === true) {
+    content.message = err.message || err || content.message;
+  }
+
+  return content;
+};
+
+module.exports = (translate, options) => {
+
+  const logger = (options || {}).logger;
```
Comment (daniel-ac-martin) [inline lib/errors.js]: Feels odd that logger is provided as a property on options but translate gets its own argument.

Is it just because the logger is optional?

File: lib/errors.js (hunk 6 above)
Comment (daniel-ac-martin) [inline lib/errors.js]: Perhaps we should provide our own very tiny translate lib that just converts the keys to English? That way, the translate would be optional like the logger.

File: lib/errors.js (hunk 5 above)
Comment (easternbloc) [inline lib/errors.js]: > `let` is for things that are intended to be scoped to a block, not a function

Comment (JoeChapman) [inline lib/errors.js]: True, but content begins as an empty object that will of course be augmented based on the outcome of the following conditions, a very typical JS pattern, and a `const` would not be allowed.

File: lib/errors.js (hunk 7 above)
Comment (JoeChapman) [inline lib/errors.js]: That was the idea, to keep the logger optional.

Comment (JoeChapman) [inline lib/errors.js]: It would be possible to make the translate function optional too, only invoke it if defined.
```
content.title = content.title || (translate && translate('errors.default.title')) || '';
```

Comment (easternbloc) [inline lib/errors.js]: 👏

File: lib/errors.js (hunk 5 above)
Comment (JoeChapman) [inline lib/errors.js]: `let` makes it possible to scope to a block that isn't a function. You know what I mean.
is `var` a thing of the past?

File: lib/errors.js (hunk 7 above)
Comment (JoeChapman) [inline lib/errors.js]: You likey?

File: lib/errors.js (hunk 5 above)
Comment (daniel-ac-martin) [inline lib/errors.js]: They avoid it in good Scala code. (But most Scala code isn't good.) Haskell makes you jump through hoops if you want mutation.
That said, some algorithms and data structures cannot be implemented efficiently without mutation.

File: lib/errors.js (hunk 7 above)
Comment (daniel-ac-martin) [inline lib/errors.js]: Change '' to 'Error' or similar?

Comment (JoeChapman) [inline lib/errors.js]: "Unknown Error'? @easternbloc @daniel-ac-martin

Comment (JoeChapman) [inline lib/errors.js]: Also, if the translate function is now optional, it should be defined as a property of the options argument, no?

**Comment by JoeChapman** [comment]
@UKHomeOffice/hof

File: lib/errors.js (hunk 7 above)
Comment (daniel-ac-martin) [inline lib/errors.js]: I'd say so.

**Comment by JoeChapman** [comment]
@daniel-ac-martin @easternbloc @joefitter can this be merged?

File: lib/errors.js [hunk 8]
```diff
@@ -0,0 +1,58 @@
+'use strict';
+
+const errorTitle = (code) => `${code}_ERROR`;
+const errorMsg = (code) => `There is a ${code}_ERROR`;
+
+const getContent = (err, debug, translate) => {
+  let content = {};
```
Comment (joefitter) [inline lib/errors.js]: I'd use `const` here as content doesn't change. Generally I would always use `const` unless you need to mutate

File: lib/errors.js [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+  const logger = (options || {}).logger;
+  const translate = (options || {}).translate;
+  const debug = (options || {}).debug;
+
+  /* eslint no-unused-vars:0 */
+  return (err, req, res, next) => {
+    /* eslint no-unused-vars:1 */
+
+    let content = getContent(err, debug, translate);
+
```
Comment (joefitter) [inline lib/errors.js]: and here

File: lib/errors.js [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+  const logger = (options || {}).logger;
+  const translate = (options || {}).translate;
+  const debug = (options || {}).debug;
+
+  /* eslint no-unused-vars:0 */
+  return (err, req, res, next) => {
+    /* eslint no-unused-vars:1 */
+
+    let content = getContent(err, debug, translate);
```
Comment (joefitter) [inline lib/errors.js]: and here

File: lib/errors.js [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+    if (logger && logger.error) {
+      logger.error(err.message || err.error, err);
+    }
+
+    res.status(err.status || 500).render('error', {
+      error: err,
+      content: debug === true ? err : content,
+      /* eslint no-process-env:0 */
```
Comment (joefitter) [inline lib/errors.js]: do we need to turn linting off here? process.env isn't used

File: test/lib/errors.js [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+
+const httpMock = require('node-mocks-http');
+
+describe('errors', () => {
+  var middleware;
+  var err;
+  var req;
+  var res;
+  var next;
+  var translate;
```
Comment (joefitter) [inline test/lib/errors.js]: `let`? or are we happy `var`ing in tests?

File: test/lib/errors.js [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+          startLink: 'my-hof-journey'
+        });
+      });
+
+    });
+  });
+
+  describe('with a logger', () => {
+
+    var logger = {};
```
Comment (joefitter) [inline test/lib/errors.js]: `const`

File: test/lib/errors.js [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+          startLink: 'my-hof-journey'
+        });
+      });
+    });
+
+  });
+
+  describe('defaults to debug false', () => {
+
+    var logger = {};
```
Comment (joefitter) [inline test/lib/errors.js]: `let` or `const` in outer scope, losing previous logger?

File: lib/errors.js (hunk 8 above)
Comment (JoeChapman) [inline lib/errors.js]: But the content of the `{}` will change, isn't that the same as `content` changing?

Comment (joefitter) [inline lib/errors.js]: no, you are not changing the assignment, properties can be added/amended/deleted on the fly, it is still the same object

However if you were to do `let foo = 'bar'; foo += 'baz';` this would error as you are altering the assignment `foo = foo + 'baz'`

Comment (joefitter) [inline lib/errors.js]: However you could do something like `const foo = 'bar'; foo.baz = 'baz';`

**Comment by JoeChapman** [comment]
@daniel-ac-martin @easternbloc @joefitter can this be merged?

**Comment by joefitter** [comment]
LGTM :shipit:

**Comment by JoeChapman** [comment]
Ok, well we need two people to verify it so @easternbloc @daniel-ac-martin are either of you willing to make a positive noise about this pr?

File: lib/cookies.js [hunk 15]
```diff
@@ -9,11 +9,10 @@ module.exports = (options) => {
     if (req.cookies && Object.keys(req.cookies).length) {
       next();
     } else if (req.query[paramName] !== undefined) {
-      var err = new Error('Cookies required');
```
Comment (daniel-ac-martin) [inline lib/cookies.js]: Arguably this should be in another PR of course.
But it's so minor does anyone really care?

**Comment by easternbloc** [comment]
@JoeChapman the only words I can muster are LGTM

**Comment by joefitter** [comment]
https://media.giphy.com/media/143vPc6b08locw/giphy.gif

File: lib/errors.js [hunk 16]
```diff
@@ -0,0 +1,57 @@
+'use strict';
+
+const errorTitle = (code) => `${code}_ERROR`;
+const errorMsg = (code) => `There is a ${code}_ERROR`;
```
Comment (daniel-ac-martin) [inline lib/errors.js]: Are we going to make this part of the official coding style?
i.e. Ban syntax like:
```js
function errorMsg(code) {
  return code + '_ERROR';
}
```

Also, how do people feel about the backtick + dollar syntax? Reminds me of PHP. I think I prefer explicit concatenation but "meh".
I suppose I just think we should be careful about using features just because they are new in ES6. I'm sure not all of them will end up in the next edition of "The good parts"! ;-)

Comment (easternbloc) [inline lib/errors.js]: tbf I tend to agree with some of this but right now we're not going against any style guides. Perhaps some of this warrants futher discussion. How about we do it at the regular Thursday gathering?

Comment (joefitter) [inline lib/errors.js]: string interpolation IMO is one of the best features in ES6! I get that it may be slightly unreadable for someone unfamiliar with the syntax, especially when used with an arrow function and implicit return, but I'd hope most users would be able see what's happening

Comment (joefitter) [inline lib/errors.js]: I'd actually argue for moving the braces around the single argument
```
const errorMsg = code => `There is a ${code}_ERROR`;
```

However, I won't mention dropping the semi colon just yet...

File: lib/errors.js [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+
+const getContent = (err, debug, translate) => {
+
+  const content = {};
+
+  if (err.code === 'SESSION_TIMEOUT') {
+    content.title = (translate && translate('errors.session.title'));
```
Comment (daniel-ac-martin) [inline lib/errors.js]: Extra parens here surely? I'm surprised the linter didn't pick them up.

Comment (easternbloc) [inline lib/errors.js]: Oh, oh, how did the linter not spot that?

File: lib/errors.js [hunk 18]
```diff
@@ -0,0 +1,57 @@
+'use strict';
+
+const errorTitle = (code) => `${code}_ERROR`;
+const errorMsg = (code) => `There is a ${code}_ERROR`;
+
+const getContent = (err, debug, translate) => {
+
+  const content = {};
```
Comment (daniel-ac-martin) [inline lib/errors.js]: Doesn't really matter for this PR. But there is probably question here about our house style and whether we want to bother making objects const. If it just makes the reference const but not the object itself I wonder if it could be misleading. Does anyone know if it is possible to have a truly immutable object in ES6?

It feels weird to me to have const all over the place and then have a load of mutation going on in objects because ES6 doesn't stop it.

Could we set "content" once and then leave it alone? Is that feasible? Does anyone but me care? ;-)

Comment (joefitter) [inline lib/errors.js]: Const isn't to do with immutability, it is to do with assignment - if you assign a value to a const you cannot reassign the const to a different value. Immutability is another thing entirely and not applicable in this example. We could use immutable.js if we want to make items immutable but that would be another discussion

File: lib/errors.js [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  return content;
+};
+
+module.exports = (options) => {
+
+  const logger = (options || {}).logger;
+  const translate = (options || {}).translate;
+  const debug = (options || {}).debug;
```
Comment (daniel-ac-martin) [inline lib/errors.js]: Wouldn't it be neater to have something like:
```js
const opts = options || {};
const logger = opts.logger;
const translate = opts.translate;
const debug = opts.debug;
```
?

Comment (JoeChapman) [inline lib/errors.js]: nah

Comment (joefitter) [inline lib/errors.js]: What would be neater would be to use default arguments and object destructuring, when node catches up :)

```js
module.exports = (options = {}) {
  const { logger, translate, debug } = options;
}
```

Comment (joefitter) [inline lib/errors.js]: but yes, your solution is neater than constantly evaluating options
